### PR TITLE
[yul] Fix sub objects data access

### LINKED ISSUE #9470: [yul] ICE when accessing a Yul data subobject: `Assembly object not found or does not contain code`
## Description

```
object "t" {
	code {
		datacopy(not(datasize("object2.object3.object4.datablock")), 0, 0)
	}
	object "object2" {
		code{}
		object "object3" {
			code{}
			object "object4" {
				code{}
				data "datablock" ""
			}
		}
	}
}
```

throws

https://github.com/ethereum/solidity/blob/4a478f0fb2d8d4f7bc7fdacfeb20e955d5396ce8/libyul/Object.cpp#L110

However

```
object "t" {
        code {
                datacopy(not(datasize("datablock")), 0, 0)
        }
        data "datablock" "okay"
}
```

compiles fine. So it seems data identifier is a valid identifier for `datasize/offset` opcode. However, nested data identifiers are not? Perhaps this is under specified?

## Environment

- Compiler version: latest develop

## Steps to Reproduce

```
$ solc --strict-assembly test.yul
```

### PR DESCRIPTION
**Pull request by 0xGeorgii**
Fixes #9470, based on #13534 

cc: @cameel

### CODE REVIEW

File: libyul/Object.cpp [hunk 1]
```diff
@@ -134,12 +134,22 @@ vector<size_t> Object::pathToSubObject(YulString _qualifiedName) const
...
+			yulAssert(object->subId != numeric_limits<size_t>::max());
+			path.push_back({object->subId});
+		}
+		else
+		{
+			yulAssert(
+ 				dynamic_cast<Data const*>(objectNode),
+ 				"Assembly object <" + _qualifiedName.str() + "> of unrecognized type."
+			);
+		}
```
Comment (nikola-matic) [inline libyul/Object.cpp]:
```suggestion
		else
			yulAssert(
				dynamic_cast<Data const*>(objectNode),
				"Assembly object <" + _qualifiedName.str() + "> of unrecognized type."
			);
```

File: libyul/Object.cpp [hunk 2]
```diff
@@ -134,12 +134,20 @@ vector<size_t> Object::pathToSubObject(YulString _qualifiedName) const
...
-		path.push_back({object->subId});
+		auto const* objectNode = object->subObjects[subIndexIt->second].get();
+		object = dynamic_cast<Object const*>(objectNode);
+		if (object)
+		{
+			yulAssert(object->subId != numeric_limits<size_t>::max());
+			path.push_back({object->subId});
+		}
+		else
+			yulAssert(
```
Comment (chriseth) [inline libyul/Object.cpp]: This should at least assert that we are in the last component of the path somehow, or that `object` is not `nullptr` at the beginning of the next iteration.

Comment (0xGeorgii) [reply libyul/Object.cpp]: > This should at least assert that we are in the last component of the path somehow, or that `object` is not `nullptr` at the beginning of the next iteration.

@chriseth  sure I will add it, but please give an advice of how such expression might look like. If write something like

`ObjectA.DataObjectA.ObjectB` it causes an error during the parsing:

```
Error parsing source.
Error: Unknown data object "ObjectA.DataObjectA.ObjectB".
 --> source:7:22:
  |
7 |         pop(datasize("ObjectA.DataObjectA.ObjectB"))
  |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
If it somehow can happen I think it makes sense to add an appropriate test.

File: libyul/Object.cpp [hunk 3]
```diff
@@ -134,12 +134,20 @@ vector<size_t> Object::pathToSubObject(YulString _qualifiedName) const
...
-		object = dynamic_cast<Object const*>(object->subObjects[subIndexIt->second].get());
-		yulAssert(object, "Assembly object <" + _qualifiedName.str() + "> not found or does not contain code.");
-		yulAssert(object->subId != numeric_limits<size_t>::max(), "");
-		path.push_back({object->subId});
+		auto const* objectNode = object->subObjects[subIndexIt->second].get();
+		object = dynamic_cast<Object const*>(objectNode);
+		if (object)
+		{
+			yulAssert(object->subId != numeric_limits<size_t>::max());
+			path.push_back({object->subId});
```
Comment (nikola-matic) [inline libyul/Object.cpp]:
```suggestion
			path.emplace_back(object->subId);
```

File: libyul/Object.h [hunk 4]
```diff
@@ -110,7 +110,7 @@ struct Object: public ObjectNode
 	/// pathToSubObject("A1.E2.F3.H4") == {1, 0, 2}
 	/// pathToSubObject("E2.F3.H4") == {1, 0, 2}
 	/// pathToSubObject("A1.E2") == {1}
-	/// The path must not lead to a @a Data object (will throw in that case).
+	/// In case the path leads to a @a Data object, returns path to its parent object.
 	std::vector<size_t> pathToSubObject(YulString _qualifiedName) const;
```
Comment (cameel) [inline libyul/Object.h]: Like I said in https://github.com/ethereum/solidity/pull/13534#discussion_r1008088129, I don't think this is the right behavior. Unless the function is only supposed to return the object that contains data and it's the caller that extracts data from that object. Is that the case? I can't see that well in tests output right now but it should be clear after the test modifications I requested in the other comment.

File: test/libyul/objectCompiler/subobject_data_access.yul [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        sstore(0, datasize("A"))
+        sstore(1, datasize("dataA1"))
+        sstore(2, datasize("dataA2"))
+        pop(dataoffset("B"))
+        pop(datasize("B.dataB1"))
+        sstore(3, calldataload(0))
+        pop(dataoffset("B.C"))
+        pop(datasize("B.C.dataC1"))
+        sstore(4, calldataload(0))
+    }
```
Comment (cameel) [inline test/libyul/objectCompiler/subobject_data_access.yul]: When it's so irregular and bunched up it's a bit hard to verify the output. The first thing I'd like to see is for example that something like `A.dataA1` gives the same size as `dataA1` used inside `A`. Another is that `B.dataB1` does not have the same size as `B`. I can't see these things here. `calldataload()` also does not seem important. My suggestion would be to split it into 3 simpler tests:

- Qualified vs unqualified data access
    ```yul
    object "A" {
        code {
            sstore(0, datasize("B"))
            sstore(1, datasize("B.data1"))
            sstore(2, datasize("B.data2"))
        }

        object "B" {
            code {
                sstore(0, datasize("B"))
                sstore(1, datasize("data1"))
                sstore(2, datasize("data2"))
            }

            data "data1" "abcdef"
            data "data2" "123"
        }
    }
    ```
- Qualified access inside the same object
    ```yul
    object "A" {
        code {
            sstore(0, datasize("A"))
            sstore(1, datasize("A.data1"))
            sstore(2, datasize("A.data2"))
        }

        data "data1" "abcdef"
        data "data2" "123"
    }
    ```
- Something more nested (like your `A.B.C` example), may include `dataoffset()`, etc.

File: libyul/Object.cpp (hunk 2 above)
Comment (cameel) [reply libyul/Object.cpp]: It can't happen, which is why it should be an assert in the code, not a test. But the only reason it can't happen is that [we have a validation against it elsewhere](https://github.com/ethereum/solidity/blob/10f81425a375d4d2622e8aa22ec7a9fb6dec30e6/libyul/AsmAnalysis.cpp#L393-L401). If you called the function directly with `ObjectA.DataObjectA.ObjectB`, it would crash because it continues the loop and tries to use `->` on a null `object`. There should be an assert at the beginning of the loop checking that `object` is not null, as @chriseth suggested, so that if anyone in the future tries to use that function without realizing that a separate validation is needed, the assert will fail instead of letting the function access memory via a null pointer.

But actually, I see we don't have a test against `ObjectA.DataObjectA.ObjectB` in `test/libyul/yulSyntaxTests/objects/` so please add one too. But that's to test the validation, not the function. The assert is needed regardless of this.

Would be a good idea to check if all important corner cases we could run into with nested data that are invalid are covered - and if not, add tests for them.

Another assert checking that we're at the last component might be useful too, but that's more tricky. We'd have to add an index to the loop or compare the length of the resulting `path` against `subObjectPathComponents`. I think that checking if `object` is not null is enough and will cover the same situations anyway.

**Review by cameel — CHANGES_REQUESTED** [review]

File: libyul/Object.cpp (hunk 2 above)
Comment (0xGeorgii) [reply libyul/Object.cpp]: @cameel thank you very much for the detailed explanation, appreciated. This is what I looked for. Will update this PR.

File: test/libyul/objectCompiler/subobject_data_access.yul (hunk 5 above)
Comment (0xGeorgii) [reply test/libyul/objectCompiler/subobject_data_access.yul]: @cameel sorry for the delay on this PR. I have a question regarding the second test you have exampled.
```
object "A" {
    code {
        sstore(0, datasize("A"))
        sstore(1, datasize("A.data1"))
        sstore(2, datasize("A.data2"))
    }

        data "data1" "abcdef"
        data "data2" "123"
}
```

In the current behavior this test wouldn't be compiled because here

https://github.com/ethereum/solidity/blob/591df042115c6df190faa26a1fb87617f7772db3/libyul/ObjectParser.cpp#L215

when data objects are being parsed it considers exact names `data1` `data2` etc.

however, when this data object is being tried to get from parsed object according to the `funCall.arguments` in

https://github.com/ethereum/solidity/blob/591df042115c6df190faa26a1fb87617f7772db3/libyul/AsmAnalysis.cpp#L395

the condition fails because of hash mismatch.

So my question is whether this case should be handled or this test should not work?
If the answer is the first option how about checking if an argument start with `this` object name and ignore this suffix, or when an object is being parsed add refs to data objects for both cases: qualified and unqualified calls. For chunks (nested objects everything works good).

Comment (cameel) [reply test/libyul/objectCompiler/subobject_data_access.yul]: Oh, so the fact that accessing the object by its name is just a special case? I did not expect that.

This case is not actually listed in the bug (and it does not trigger an ICE) so we could do it either way. On one hand it seems that this is a pretty natural extension of the syntax and should work. On the other, I now see that this would result in objects having more than one path. For that reason maybe let's not change it here - let's keep it producing an error. But please still add this case as a syntax test.

> If the answer is the first option how about checking if an argument start with `this` object name and ignore this suffix, or when an object is being parsed add refs to data objects for both cases: qualified and unqualified calls.

If we were to implement this, I think the approach with removing the prefix would be better than having duplicated entries.

But as I said, for now it's fine as an error.

Comment (0xGeorgii) [reply test/libyul/objectCompiler/subobject_data_access.yul]: Test added, #13807 was added not to forget it despite low priority and probably will not be implemented, just for history.

File: libyul/Object.cpp [hunk 6]
```diff
@@ -128,18 +128,33 @@ vector<size_t> Object::pathToSubObject(YulString _qualifiedName) const
...
+		if (object)
+		{
+			yulAssert(object->subId != numeric_limits<size_t>::max());
+			path.emplace_back(object->subId);
+		}
+		else
+		{
+			yulAssert(
+				i == subObjectPathComponents.size() - 1,
+				"Assembly object <" + _qualifiedName.str() + "> of incorrect path."
```
Comment (cameel) [inline libyul/Object.cpp]: This would be more accurate:

```suggestion
				"Assembly object <" + _qualifiedName.str() + "> does not exist. <" + path + "> has no nested objects."
```

But you'd have to use something like `joinHumanReadable()` on path to convert it to a string.

File: test/libyul/objectCompiler/nested_subobject_data_access.yul [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+//     /* "source":74:75   */
+//   0x01
+//     /* "source":67:100   */
+//   sstore
+//     /* "source":119:141   */
+//   dataSize(sub_0.sub_0)
+//     /* "source":116:117   */
+//   0x02
+//     /* "source":109:142   */
+//   sstore
```
Comment (cameel) [inline test/libyul/objectCompiler/nested_subobject_data_access.yul]: This is wrong. It means that the sizes returned for `B.C.dataC1` and `B.C.dataC2` are the same and all equal to the size of `B.C` rather than being the sizes of the data elements.

This confirms what I said in https://github.com/ethereum/solidity/pull/13707/files#r1032515177, i.e. `Object::pathToSubObject()` should not be returning the path to the parent object. It should return the path to the data object instead.

File: test/libyul/objectCompiler/nested_subobject_data_access.yul [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            sstore(1, datasize("dataB1"))
+            sstore(2, datasize("dataB2"))
+        }
+
+        data "dataB1" "abcdef"
+        data "dataB2" "123"
+
+        object "C" {
+            code {
+                sstore(0, datasize("dataC1"))
```
Comment (cameel) [inline test/libyul/objectCompiler/nested_subobject_data_access.yul]: Would not hurt to have all paths here for easy comparison with `A`. They should resolve to the same sizes in `A` and `C`.
```suggestion
                sstore(0, datasize("C"))
                sstore(0, datasize("dataC1"))
                sstore(0, datasize("dataC2"))
```

File: test/libyul/objectCompiler/nested_subobject_data_access.yul [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+    object "B" {
+        code {
+            sstore(0, datasize("B"))
+            sstore(1, datasize("dataB1"))
+            sstore(2, datasize("dataB2"))
+        }
+
+        data "dataB1" "abcdef"
+        data "dataB2" "123"
```
Comment (cameel) [inline test/libyul/objectCompiler/nested_subobject_data_access.yul]: I'd remove this part. We have it in `subobject_qualified_vs_unqualified_data_access.yul` and here it does not seem to add much.

Instead I'd add `dataoffset()` in `A` and `C`. It should work the same way `datasize()` does but would not hurt to have it covered.

File: test/libyul/objectCompiler/subobject_qualified_vs_unqualified_data_access.yul [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+//     /* "source":72:73   */
+//   0x01
+//     /* "source":65:95   */
+//   sstore
+//     /* "source":114:133   */
+//   dataSize(sub_0)
+//     /* "source":111:112   */
+//   0x02
+//     /* "source":104:134   */
+//   sstore
```
Comment (cameel) [inline test/libyul/objectCompiler/subobject_qualified_vs_unqualified_data_access.yul]: Same here. These sizes are wrong. They should be the same as for `B` below.

**Review by cameel — CHANGES_REQUESTED** [review]
Sorry for they delay in reviewing this. Here's another batch of feedback.

Basically, there's one remaining major issue and it's that the PR makes `A.B.C.data` equivalent to `A.B.C`, so returned sizes and offsets will be wrong. This needs to be fixed.

Other than that, just some test and message tweaks.

File: test/libyul/objectCompiler/subobject_data_access.yul (hunk 5 above)
Comment (cameel) [reply test/libyul/objectCompiler/subobject_data_access.yul]: @GeorgePlotnikov Can you add the second case from https://github.com/ethereum/solidity/pull/13707#discussion_r1032532033 as a syntax test if we don't have it already? I don't remember if we do, but if we don't, we should add a syntax test showing that this is not valid.

**Comment by cameel** [comment]
Also, this is missing a changelog entry.

**Comment by NunoFilipeSantos** [comment]
Hi @GeorgePlotnikov ! 👋 are you planning to come back soon to this PR, or should we close it for now?

**Comment by 0xGeorgii** [comment]
> Hi @GeorgePlotnikov ! 👋 are you planning to come back soon to this PR, or should we close it for now?

hey Nuno, excuse me for the delay, quite busy with other tings. Yes, I am chasing to complete this PR definitely. Will get into it in 2 weeks if it works.

**Comment by NunoFilipeSantos** [comment]
Hi @GeorgePlotnikov ! No problem. We were primarily interested in knowing if some external PRs have been abandoned. 😃
